mypy/build.py

```diff
@@ -2340,6 +2342,11 @@ def finish_passes(self) -> None:
         manager = self.manager
         if self.options.semantic_analysis_only:
             return
+        if manager.options.disallow_undefined_vars:
+            manager.errors.set_file(self.xpath, self.tree.fullname, options=manager.options)
+            self.tree.accept(
```

Comment:
I'm not sure if this was the right place to plug this in. It needed to be done before `self.free_state()` is called.

Reply:
I think that you could add another method, such as `detect_partially_undefined_vars` (feel free to alter the name) that performs the analysis if enabled, and call it from `process_stale_scc` after calling `type_check_second_pass` and before calling `finish_passes`. I.e., this would be an additional pass after type checking.